### DISCUSSION TITLE
Solution to Error "mined block in the past" due to windowpost and winpost happening at the same time - Solution not verified by lotus team

### DISCUSSION THREAD

**Discussion by phoenixzhua** [filecoin-project/lotus#7797] [Tutorials]
**Summary:**
1. Add a flag to identify whether it is wdPost or winPost.
2. Add an env. variable:  WINPOST_TMPDIR for the winning post.

**Build:**
Put those updated files to your cargo folder (e.g. ~/.cargo/registry/src/github.com-1ecc6299db9ec823). and build it according to the official guide. 
```
cd /home/miner/Download/lotus/
git checkout release/v1.13.0
git submodule update --init --recursive
RUSTFLAGS="-C target-cpu=native -g" FFI_BUILD_FROM_SOURCE="1" FFI_USE_GPU2=1 FFI_USE_CUDA=1 make clean all
sudo make install
```

**Run:**
If you start miner with systemctl, add a WINPOST_TMPDIR env. variable to the EnvironmentFile specified in the miner service file.
```
$ cat miner.service
...
EnvironmentFile=/etc/ipfs/env/miner.env
...

$ cat /etc/ipfs/env/miner.env
...
#GPU locked folder for wdPost
TMPDIR=/etc/ipfs/tmpDir 

#GPU locked folder for winPost, different with the TMPDIR
WINPOST_TMPDIR=/etc/ipfs/tmpDir_win 
...
```

In case you start miner with script, export the WINPOST_TMPDIR env. variable before you start the miner script. Make sure it is different from the TMPDIR env. variable
`export WINPOST_TMPDIR=/etc/ipfs/tmpDir_win`

Please let us know if you have any question.

**Comment by phoenixzhua** [comment]
Please download the updated code at the link below: 
https://ipfs.io/ipfs/QmRaZLP38Y8Jf2GEDXivHBUf3xXArBoQ833qs4PEoaktsw?filename=rust_code.zip
or
https://bafybeiei36tdae3up342hozqizolzvqi6w5zexmsm77e32aonyh3bceice.ipfs.dweb.link/

**Comment by phoenixzhua** [comment]
The fix is based on the lotus version 1.13.0: 
```
Daemon:  1.13.0+mainnet+git.7a55e8e89.dirty+api1.2.0
Local: lotus-miner version 1.13.0+mainnet+git.7a55e8e89.dirty
```

**Comment by leozhao99** [comment]
BTW, we use two RTX 3090 GPU cards on a miner box.

**Comment by RobQuistNL** [comment]
Could you open the changes in a merge request so we can easily see the diff?

**Reply by phoenixzhua** [reply]
Thank you for your reply, RobQuistNL. Those updated files is related to several repositories, so I create my repository to bring the related part of those repositories together, and merge files to my repository to show the diff.
https://github.com/phoenixzhua/cargo-files/commit/b47bf69bcde9215f30304606fc31036a0ac362cd

**Reply by RobQuistNL** [reply]
But the entire purpose of github, git and open source software is the ability to merge in improvements :P 

Then you can also have developers review and verify your changes.

**Comment by lanpin** [comment]
This is a simple and effective solution.
